generator/src/main/java/dev/cdevents/generator/PreprocessSchemas.java

```diff
@@ -0,0 +1,66 @@
+package dev.cdevents.generator;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Iterator;
+
+public class PreprocessSchemas {

```

Comment:
Can we add some documentation for this class itself? I know we have the README, but it would also be nice to document this class along with the purpose.

Reply:
Documented as part of `DEVELOPMENT.md` as this will be used in the context of generating new SDK release

Reply:
Hmm, Im kind of on the fence with this one. While I **love** the details and descriptions in the `md` file. I think having a simple overall comment about its purpose will at least people going through the code understand what it's doing.

```
/**
   * Adapts incoming schemas to a format that jsonschema2 pojo
   */
```

Also, I wonder if this class should implement an interface. If you remember last SIG we talked about potentially finding a more suitable modeling language. If we had an interface, the code wouldn't need to change much.

It's also kind of odd that this `PreprocessSchemas` class has a main. So this is intended to be ran as a CLI? Why? Im assuming the `CDEventsGenerator` calls the `jsonschema2pojo` class. Why not just add the preprocessor there? Do we really need a separate main?

In addition, if we do decide to embed into the generator, please make it implement a preprocessor interface. If we decided to ever abandon jsonschema, having an interface will make that much easier.

Reply:
yes we need separate main for `PreprocessSchemas` as this class will be executed before `jsonschema2pojo` runs to update the schemas with `.json` extension. And `CDEventsGenerator` needs to be executed once after `jsonschema2pojo` generates pojo from schema.
Discussed the same at this, https://github.com/cdevents/sdk-java/pull/82#discussion_r1611386327

Reply:
The class `PreprocessSchemas` runs as maven executed java goal, which needs `main`.
Updating PreprocessSchemas itself to have the interface which provides the flexibility in implementation.

Reply:
This needs to be refactored. It is kind of backwards. What I mean by that is the preprocessor should not ever be the entrypoint to generation. The entrypoint, or main, should be in the generation class which calls preprocessors. That is much more flexible

Reply:
CDEventsGenerator does not calls the `jsonschema2pojo` class.

The flow is something like this,
`jsonschema2pojo` is invoked directly from [generator/pom.xml](https://github.com/cdevents/sdk-java/blob/main/generator/pom.xml#L69-L588) to generate model classes 
and these model classes are used to generate various CDEvents using `CDEventsGenerator` class, which is invoked after `jsonschema2pojo` from [generator/pom.xml](https://github.com/cdevents/sdk-java/blob/main/generator/pom.xml#L592-L611)

And this `PreprocessSchemas` needs to run before `jsonschema2pojo` to update the schemas that comply with `jsonschema2pojo` for generating model classes.

I am not sure we have any other options to refactor, but I do not think we can use generation class as a single entry point to invoke preprocessors in this case?

Reply:
Actually, the more I think about this, I think the less of a big deal it is. I realized that if we switched protocols, jsonschema to openAPI for example, then this just goes away and there probably would not be anything to replace it.